src/rustc-driver.md

```diff
@@ -0,0 +1,66 @@
+# The Rustc Driver
+
+The [`rustc_driver`] is essentially `rustc`'s `main()` function. It acts as
+the glue for running the various phases of the compiler in the correct order,
+managing state such as the [`CodeMap`] \(maps AST nodes to source code),
+[`Session`] \(general build context and error messaging) and the [`TyCtxt`]
+\(the "typing context", allowing you to query the type system and other cool
+stuff). The `rustc_driver` crate also provides external users with a method
+for running code at particular times during the compilation process, allowing
+third parties to effectively use `rustc`'s internals as a library for
+analysing a crate.
+
+For those using `rustc` as a library, the `run_compiler()` function is the main
+entrypoint to the compiler. Its main parameters are a list of command-line
+arguments and a reference to something which implements the `CompilerCalls`
+trait. A `CompilerCalls` creates the overall `CompileController`, letting it
+govern which compiler passes are run and attach callbacks to be fired at the end
+of each phase.
+
+From `rustc_driver`'s perspective, the main phases of the compiler are:
+
+1. *Parse Input:* Initial crate parsing
+2. *Configure and Expand:* Resolve `#[cfg]` attributes and expand macros
+3. *Run Analysis Passes:* Run the resolution, typechecking, region checking
```

Comment:
Not sure what 'resolution' means here, but name resolution now happens earlier. I think 'borrow-checking' is more common than 'region-checking'

Reply:
I think _trait_ resolution happens here though, I think...

Reply:
yes, correct

Reply:
Haha, I'm not sure either! The [rustdocs] just say "resolution" so I thought I'd fudge it :stuck_out_tongue: 

[rustdocs]: https://michael-f-bryan.github.io/rustc-internal-docs/rustc_driver/driver/fn.phase_3_run_analysis_passes.html